helper.py

```diff
@@ -0,0 +1,214 @@
+
+    def get_date(self):
+        return self.date
+
+    def get_highest_temp(self):
+        return self.highest_temp
+
+    def get_lowest_temp(self):
+        return self.lowest_temp
+
+    def get_humidity(self):        
+        return self.humidity
+
+    def __str__(self):
+        return f"Date: {self.date}, Highest Temp: {self.highest_temp},Lowest Temp: {self.lowest_temp}, Humidity: {self.humidity}"
+
+
+# function to read files , deals with NULLs, empty columns as well
+def read_file(path):
+    data = []
+    with open(path, 'r') as file:
+        first_line = file.readline().strip().split(',')
+        if len(first_line) < 8:  # Check if the first line has enough values
+            file.readline()
+        for line in file:
+            data_values = line.strip().split(',')
+            if len(data_values) < 8:  # Check if the line has enough values
+                continue
+            date = data_values[0]
+            highest_temp = data_values[1]
+            lowest_temp = data_values[3]
+            humidity = data_values[7]
+
+            # Handle missing values by setting them to 0
+            highest_temp = int(highest_temp) if highest_temp.strip() else 0
+            lowest_temp = int(lowest_temp) if lowest_temp.strip() else 0
+            humidity = int(humidity) if humidity.strip() else 0
+
+            required_entry = Weather_man(date, highest_temp, lowest_temp, humidity)
+            data.append(required_entry)
+
+    return data
+
+# computes highest temp, lowest temp, highest humidity along years alone
+def Yearly_Computations(data):
+    if not data:
+        print("No data available for this year :(\n")
+    highest_temp = 0
+    lowest_temp = 0
+    high_humidity = 0
+    date1 = None
+    date2 = None
+    date3 = None
+    # For each entry,retrieve three values: temp1, temp2, and temp3, 
+    # which represent the highest temperature, lowest temperature, and 
+    # humidity, respectively, for that particular entry.
+    for entry in data:
+        temp1 = entry.get_highest_temp()
+        temp2 = entry.get_lowest_temp()
+        temp3 = entry.get_humidity()
+        if (highest_temp == 0) or (temp1 > highest_temp):
+            highest_temp = temp1
+            date1 = entry.get_date()
+        if (lowest_temp == 0) or (temp2 < lowest_temp):
+            lowest_temp = temp2
+            date2 = entry.get_date()
+        if (high_humidity == 0) or (temp3 > high_humidity):
+            high_humidity = temp3
+            date3 = entry.get_date()
+    return highest_temp, date1, lowest_temp, date2, high_humidity, date3
+
+
+"""
+yearly_stats() computes and prints the yearly statistics for a collection of
+data rows representing different years. It uses the Yearly_Computations(data)
+function to process each individual year's data and then aggregates the results
+to find the overall highest temperature, lowest temperature, and highest
+humidity among all the years
+"""
```

Comment:
Docstring is written at the start of the function under its definition and if you are writing it for keep in consistent and write it for all functions.

Reply:
got it